Multisig/UI/Assets/BalancesViewController/BalancesV2/BalancesTableViewController.swift

```diff
@@ -0,0 +1,202 @@
+        tableView.allowsSelection = false
+        tableView.backgroundColor = .primaryBackground
+        tableView.refreshControl = makeRefreshControl()
+
+        // register cells
+        tableView.registerCell(ImportKeyBannerCell.self)
+        tableView.registerCell(CreatePasscodeBannerCell.self)
+        tableView.registerCell(TotalCell.self)
+        tableView.registerCell(BalanceCell.self)
+
+        // auto-sizing
+        tableView.rowHeight = UITableView.automaticDimension
+        tableView.estimatedRowHeight = 60
+    }
+
+    private func makeRefreshControl() -> UIRefreshControl {
+        let control = UIRefreshControl()
+        return control
+    }
+
+    /// Updates the UI based on the new data
+    /// - Parameter newData: updated screen data
+    func reload(_ newData: Balances) {
+        data = newData
+
+        tableView.reloadData()
+
+        if data.isRefreshing {
+            tableView.refreshControl?.beginRefreshing()
+        } else if tableView.refreshControl?.isRefreshing == true {
+            tableView.refreshControl?.endRefreshing()
+        }
+    }
+
+    // MARK: UITableViewDataSource implementation
+
+    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        data.count
+    }
+
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cellData = data[indexPath]
+
+        switch cellData {
+
+        case is Balances.ImportKeyBannerPlaceholder:
+            return tableView.dequeueCell(ImportKeyBannerCell.self, for: indexPath)
+
+        case is Balances.PasscodeBannerPlaceholder:
+            return tableView.dequeueCell(CreatePasscodeBannerCell.self, for: indexPath)
+
+        case let total as Balances.Total:
+            let cell = tableView.dequeueCell(TotalCell.self, for: indexPath)
+            cell.detailText = total.quoteAmount
+            return cell
+
+        case let entry as Balances.Entry:
+            let cell = tableView.dequeueCell(BalanceCell.self, for: indexPath)
+            cell.mainText = entry.name
+            cell.detailText = entry.tokenAmount
+            cell.subDetailText = entry.quoteAmount
+            cell.image = entry.image
+            return cell
+
+        default:
+            return UITableViewCell()
+        }
+    }
+}
+
+// MARK: UI Data
+
+extension BalancesTableViewController {
+    /// UI data model for the controller
+    struct Balances {
+        static let empty = Balances()
+
+        /// Whether the refresh indicator is shown or not
+        var isRefreshing: Bool = false
+
+        /// If not nil, then a banner will be shown.
+        /// Only one banner at a time.
+        var banner: BannerPlaceholder?
+
+        /// If not nil, then the entries and total cells will be shown
+        var summary: Summary?
+
+        /// Creates cell data from the banner and summary
+        var cellData: [BalancesCellData] {
```

Comment:
as this is an array I would make name plural. `cellsData`

Reply:
🤔 but "data" is already a plural